misc-scripts/xref_mapping/XrefParser/MGIParser.pm

```diff
@@ -23,82 +23,57 @@ use strict;
     return 1;    # 1 is an error
   }
-  my %label;
-  my %version;
-  my %description;
-  my %accession;
-
-  my $sql = 'select source_id from source where name like "MGI" and priority_description like "descriptions"';
-  my $sth = $dbi->prepare($sql);
-  
-  $sth->execute();
-  my ($mgi_source_id);
-  $sth->bind_columns(\$mgi_source_id);
-  my @arr;
-  while($sth->fetch()){
-    push @arr, $mgi_source_id;
-  }
-  $sth->finish;
-  
-  $sql = "select accession, label, version,  description from xref where source_id in (".join(", ",@arr).")";
-
-  $sth = $dbi->prepare($sql);
-  $sth->execute();
-  my ($acc, $lab, $ver, $desc);
-  $sth->bind_columns(\$acc, \$lab, \$ver, \$desc);
-  while (my @row = $sth->fetchrow_array()) {
-    if(defined($desc)){
-      $accession{$lab} = $acc;
-      $label{$acc} = $lab;
-      $version{$acc} = $ver;
-      $description{$acc} = $desc;
-    }
-  }
-  $sth->finish;
 
   #synonyms
-  $sql = "insert ignore into synonym (xref_id, synonym) values (?, ?)";
+  my $sql = "insert ignore into synonym (xref_id, synonym) values (?, ?)";
   my $add_syn_sth = $dbi->prepare($sql);    
-
   my $syn_hash = $self->get_ext_synonyms("MGI", $dbi);
 
+  #Init input file
+  my $input_file = Text::CSV->new({
+                                    sep_char           => "\t",
+				    empty_is_undef     => 1,
+                                    strict => 1,
+                                    allow_loose_quotes => 1,
+				   }) or croak "Cannot use file $file: ".Text::CSV->error_diag ();
+
+  # init headers
+  # MGI:1915941	1110028C15Rik	RIKEN cDNA 1110028C15 gene	33.61	1	ENSMUSG00000026004	ENSMUST00000042389 ENSMUST00000068168 ENSMUST00000113987 ENSMUST00000129190 ENSMUST00000132960	ENSMUSP00000036975 ENSMUSP00000063843 ENSMUSP00000109620 ENSMUSP00000118603
+  $input_file->column_names([qw(accession symbol name position chrom ens_gene_stableid)] ); #ignore last two columns EnsemblTranscriptIDs and EnsemblProteinIDs
   my $count = 0;
   my $syn_count = 0;
-  while ( my $line = $file_io->getline() ) {
-#MGI:1915941	1110028C15Rik	RIKEN cDNA 1110028C15 gene	33.61	1	ENSMUSG00000026004	ENSMUST00000042389 ENSMUST00000068168 ENSMUST00000113987 ENSMUST00000129190 ENSMUST00000132960	ENSMUSP00000036975 ENSMUSP00000063843 ENSMUSP00000109620 ENSMUSP00000118603
+  while ( my $data = $input_file->getline_hr( $file_io ) ){
 
-    if($line =~ /(MGI:\d+).*(ENSMUSG\d+)/){
-      my $acc = $1;
-      my $ensid = $2;
+      my $acc = $data->{'accession'};
+      my $ensid = $data->{'ens_gene_stableid'};
       my $xref_id = $self->add_xref({ acc        => $acc,
-				      version    => $version{$acc},
-				      label      => $label{$acc},
-				      desc       => $description{$acc},
+				      version    => 0,
+				      label      => $data->{'symbol'},
+				      desc       => $data->{'name'},
 				      source_id  => $source_id,
                                       dbi        => $dbi,
 				      species_id => $species_id,
```

Comment:
Another "GitHub will not let me attach a comment to the correct line": the _add_direct_xref()_ call below inserts an empty string into the database. Please change the fourth argument to _undef_.